src/StructuredLogger.Utils/SecretsSearch.cs

```diff
@@ -0,0 +1,130 @@
+                var foundResults = ScanForSecrets(_build.StringTable.Instances, activeDetectors, maxResults);
+
+                foreach (var result in foundResults)
+                {
+                    results.Add(result);
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        public List<SecretDescriptor> SearchSecrets(string text, NodeQueryMatcher? matcher, int maxResults)
+        {
+            var activeDetectors = GetActiveDetectors(matcher == null ? [] : [matcher]);
+
+            var secrets = DetectSecrets(text, activeDetectors);
+
+            return secrets.Take(maxResults).ToList();
+        }
+
+        private IEnumerable<SearchResult> ScanForSecrets(IEnumerable<string> stringsPool, Dictionary<SensitiveDataKind, ISensitiveDataDetector> detectors, int maxResults)
+        {
+            var secretsSet = new HashSet<string>();
+            foreach (var text in stringsPool)
+            {
+                var secretResults = DetectSecrets(text, detectors);
+                foreach (SecretDescriptor secretDescriptor in secretResults)
+                {
+                    secretsSet.Add(secretDescriptor.Secret);
+                }
+            }
+
+            var results = new List<SearchResult>();
+            if (_build.SearchIndex is { } index)
+            {
+                foreach (var text in secretsSet)
+                {
+                    index.MaxResults = maxResults;
+                    index.MarkResultsInTree = false;
+                    IEnumerable<SearchResult> indexResults = index.FindNodes(text, CancellationToken.None);
+                    if (indexResults.Any())
+                    {
+                        results.AddRange(indexResults);
+                    }
+                }
+            }
+
+            return results;
+        }
+
+        private List<SecretDescriptor> DetectSecrets(string text, Dictionary<SensitiveDataKind, ISensitiveDataDetector> detectors)
+        {
+            if (_secretCache.TryGetValue(text, out var cachedSecrets))
+            {
+                return cachedSecrets
+                    .Where(kv => detectors.Any(d => d.Key == kv.Key))
+                    .SelectMany(kv => kv.Value)
+                    .ToList();
+            }
+
+            var results = new Dictionary<SensitiveDataKind, List<SecretDescriptor>>();
+
+            foreach (var detector in detectors)
+            {
+                Dictionary<SensitiveDataKind, List<SecretDescriptor>> detectedSecrets = detector.Value.Detect(text);
+                foreach (KeyValuePair<SensitiveDataKind, List<SecretDescriptor>> kv in detectedSecrets)
+                {
+                    if (kv.Value.Any())
+                    {
+                        results[kv.Key] = kv.Value;
+                    }
+                }
+            }
+
+            if (results.Any())
+            {
+                _secretCache[text] = results;

```

Comment:
Is this guaranteed to be called single-threaded from the UI?

Reply:
no idea, @KirillOsenkov could you advice?

Reply:
It is called on a background thread, but there's no parallelism involved, so no need for a lock. When in doubt you can put a breakpoint in the code and see from which thread it is called:

![image](https://github.com/user-attachments/assets/d27b25e1-f299-46cf-9b6d-b308a52988cd)